File: scala/ch02/ex2.2.scala
type point = Tuple2[Double,Double]

def makePoint (x: Double, y: Double) = (x, y)

def xPoint (p: point) = p._1
def yPoint (p: point) = p._2
  
type segment = Tuple2[point,point]

def makeSegment(start: point, end: point) = (start, end)
  
def startSegment (s: segment) = s._1
def endSegment   (s: segment) = s._2

def midpointSegment (s: segment) = {
  val x1 = xPoint (startSegment(s))
  val y1 = yPoint (startSegment(s))
  val x2 = xPoint (endSegment(s))
  val y2 = yPoint (endSegment(s))
  makePoint ((x1 + x2) / 2, (y1 + y2) / 2)
}

def printPoint (p: point) = format ("(%f,%f)\n", xPoint(p), yPoint(p))

val zeroZero = makePoint (0.0, 0.0)
val zeroTwo  = makePoint (0.0, 2.0)
val twoZero  = makePoint (2.0, 0.0)
val zeroOne  = makePoint (0.0, 1.0)
val oneZero  = makePoint (1.0, 0.0)
val oneOne   = makePoint (1.0, 1.0)
val twoTwo   = makePoint (2.0, 2.0)
val mOneMOne = makePoint (-1.0, -1.0)
val mTwoMTwo = makePoint (-2.0, -2.0)

val zeroZeroZeroZero = makeSegment (zeroZero, zeroZero)
val zeroZeroTwoTwo   = makeSegment (zeroZero, twoTwo)
val zeroZeroMOneMOne = makeSegment (zeroZero, mOneMOne)
val zeroZeroMTwoMTwo = makeSegment (zeroZero, mTwoMTwo)
val zeroZeroTwoZero  = makeSegment (zeroZero, twoZero)
val zeroZeroZeroTwo  = makeSegment (zeroZero, zeroTwo)
val twoTwoZeroZero   = makeSegment (twoTwo, zeroZero)
val mTwoMTwoZeroZero = makeSegment (mTwoMTwo, zeroZero)
val twoZeroZeroZero  = makeSegment (twoZero, zeroZero)
val zeroTwoZeroZero  = makeSegment (zeroTwo, zeroZero)
val mTwoMTwoTwoTwo   = makeSegment (mTwoMTwo, twoTwo)
val twoTwoMTwoMTwo   = makeSegment (twoTwo, mTwoMTwo)

val midZeroZeroA = midpointSegment (zeroZeroZeroZero)
val midZeroZeroB = midpointSegment (mTwoMTwoTwoTwo)
val midZeroZeroC = midpointSegment (twoTwoMTwoMTwo)
val midOneOne    = midpointSegment (zeroZeroTwoTwo)
val midOneZeroA  = midpointSegment (zeroZeroTwoZero)
val midOneZeroB  = midpointSegment (twoZeroZeroZero)
val midZeroOneA  = midpointSegment (zeroZeroZeroTwo)
val midZeroOneB  = midpointSegment (zeroTwoZeroZero)

import org.scalatest._
import org.scalatest.matchers._

object segmentsSpec extends Spec with ShouldMatchers {
  def isEqualPoint(p :point, expected: point) = {
    xPoint(p) should equal (xPoint(expected))
    yPoint(p) should equal (yPoint(expected))
  }
  
  describe ("midpoint calculation of Segments") {
    it ("should find the point in the middle") {
      isEqualPoint (midZeroZeroA, zeroZero)
      isEqualPoint (midZeroZeroB, zeroZero)
      isEqualPoint (midZeroZeroC, zeroZero)
      isEqualPoint (midOneOne,    oneOne)
      isEqualPoint (midOneZeroA,  oneZero)
      isEqualPoint (midOneZeroB,  oneZero)
      isEqualPoint (midZeroOneA,  zeroOne)
      isEqualPoint (midZeroOneB,  zeroOne)
    }
  }
}
segmentsSpec execute

printPoint (midZeroZeroA)
printPoint (midZeroZeroB)
printPoint (midZeroZeroC)
printPoint (midOneOne)
printPoint (midOneZeroA)
printPoint (midOneZeroB)
printPoint (midZeroOneA)
printPoint (midZeroOneB)
println("")
printPoint (zeroZero)
printPoint (oneZero)
printPoint (zeroOne)
printPoint (twoZero)
printPoint (zeroTwo)
printPoint (oneOne)
printPoint (twoTwo)
printPoint (mOneMOne)
printPoint (mTwoMTwo)
println("")

File: clojure/ch02/ex2.2.clj
(ns sicp.ch02 (:use clojure.contrib.test-is))

(defn make-point [x y] [x y])

(defn x-point [p] (get p 0))
(defn y-point [p] (get p 1))

(defn make-segment [start end] [start end])
  
(defn start-segment [s] (get s 0))
(defn end-segment [s]   (get s 1))

(defn midpoint-segment [s]
  (let [x1 (x-point (start-segment s))
        y1 (y-point (start-segment s))
        x2 (x-point (end-segment s))
        y2 (y-point (end-segment s))]
      (make-point (/ (+ x1 x2) 2) (/ (+ y1 y2) 2))))

(defn print-point [p] (println (format "(%f,%f)" (x-point p) (y-point p))))

(def zero-zero (make-point 0.0 0.0))
(def zero-two  (make-point 0.0 2.0))
(def two-zero  (make-point 2.0 0.0))
(def zero-one  (make-point 0.0 1.0))
(def one-zero  (make-point 1.0 0.0))
(def one-one   (make-point 1.0 1.0))
(def two-two   (make-point 2.0 2.0))
(def m-one-m-one (make-point -1.0 -1.0))
(def m-two-m-two (make-point -2.0 -2.0))

(def zero-zero-zero-zero   (make-segment zero-zero zero-zero))
(def zero-zero-two-two     (make-segment zero-zero two-two))
(def zero-zero-m-one-m-one (make-segment zero-zero m-one-m-one))
(def zero-zero-m-two-m-two (make-segment zero-zero m-two-m-two))
(def zero-zero-two-zero    (make-segment zero-zero two-zero))
(def zero-zero-zero-two    (make-segment zero-zero zero-two))
(def two-two-zero-zero     (make-segment two-two zero-zero))
(def m-two-m-two-zero-zero (make-segment m-two-m-two zero-zero))
(def two-zero-zero-zero    (make-segment two-zero zero-zero))
(def zero-two-zero-zero    (make-segment zero-two zero-zero))
(def m-two-m-two-two-two   (make-segment m-two-m-two two-two))
(def two-two-m-two-m-two   (make-segment two-two m-two-m-two))

(def mid-zero-zero-a (midpoint-segment zero-zero-zero-zero))
(def mid-zero-zero-b (midpoint-segment m-two-m-two-two-two))
(def mid-zero-zero-c (midpoint-segment two-two-m-two-m-two))
(def mid-one-one     (midpoint-segment zero-zero-two-two))
(def mid-one-zero-a  (midpoint-segment zero-zero-two-zero))
(def mid-one-zero-b  (midpoint-segment two-zero-zero-zero))
(def mid-zero-one-a  (midpoint-segment zero-zero-zero-two))
(def mid-zero-one-b  (midpoint-segment zero-two-zero-zero))

(deftest test-segments
  (defn is-equal-point? [p expected-p]
    (is (= (x-point p) (x-point expected-p)))
    (is (= (y-point p) (y-point expected-p))))
  
  (is-equal-point? mid-zero-zero-a zero-zero)
  (is-equal-point? mid-zero-zero-b zero-zero)
  (is-equal-point? mid-zero-zero-c zero-zero)
  (is-equal-point? mid-one-one     one-one)
  (is-equal-point? mid-one-zero-a  one-zero)
  (is-equal-point? mid-one-zero-b  one-zero)
  (is-equal-point? mid-zero-one-a  zero-one)
  (is-equal-point? mid-zero-one-b  zero-one))

(run-tests)

(print-point mid-zero-zero-a)
(print-point mid-zero-zero-b)
(print-point mid-zero-zero-c)
(print-point mid-one-one)
(print-point mid-one-zero-a)
(print-point mid-one-zero-b)
(print-point mid-zero-one-a)
(print-point mid-zero-one-b)
(println "")
(print-point zero-zero)
(print-point one-zero)
(print-point zero-one)
(print-point two-zero)
(print-point zero-two)
(print-point one-one)
(print-point two-two)
(print-point m-one-m-one)
(print-point m-two-m-two)
(println "")
